### ISSUE TITLE
New buzzwords The PredictionIO Machine Learning Project Propels Intelligent Development

### ISSUE THREAD

**Issue by sudocurse** [sudocurse/reading-list#2296]
The PredictionIO Machine Learning Project Propels Intelligent Development<br>
http://ift.tt/2CM6OsH<br>
<img width="640" height="426" src="http://ift.tt/2oVTIHh" class="webfeedsFeaturedVisual wp-post-image wp-stateless-item" style="display: block; margin: auto; margin-bottom: 20px;" data-image-size="large" data-stateless-media-bucket="cdn.thenewstack.io" data-stateless-media-name="media/2017/11/0d9cf7ca-magic8ball.jpg">
<p>PredictionIO, one of the latest Apache Software Foundation projects elevated to <a href="http://ift.tt/2zyp0Vh" class="ext-link" rel="external nofollow">top-level</a> status, like many such projects, grew out of frustration.</p>
<p>&ldquo;Every time we needed to build something intelligent &mdash; doing a news recommendation or resume analysis, product matching, something like that &mdash; we had to build the machine learning tech stack from scratch every time,&rdquo; said <a href="https://twitter.com/simonchannet" class="ext-link" rel="external nofollow">Simon Chan</a>, co-founder of <a href="http://ift.tt/2y328AR" class="ext-link" rel="external nofollow">PredictionIO</a>, along with&nbsp;<a href="https://twitter.com/kpszeto?lang=en" class="ext-link" rel="external nofollow">Donald Szeto</a>,&nbsp;<a href="https://twitter.com/ckhkenneth" class="ext-link" rel="external nofollow">Kenneth Chan&nbsp;</a>and&nbsp;<a href="http://ift.tt/2BLjrHV" class="ext-link" rel="external nofollow">Thomas Stone</a>.</p>
<p>&ldquo;If you don&rsquo;t have to build your database from scratch, why should you have to build your machine learning server from scratch? &hellip; The idea was to build something developers could use without a Ph.D.&rdquo;</p>
<p>They were inspired by the model of MySQL at that time, he said, and determined to offer an open source version of it.</p>
<p>Salesforce <a href="http://ift.tt/2bTgdBU" class="ext-link" rel="external nofollow">acquired</a> their company in early 2016 and donated an open source version to ASF later that year. It&rsquo;s since become the backbone of Salesforce <a href="http://ift.tt/2sTEa3N" class="local-link">Einstein</a>, which enables customers to build their own AI-powered tools.</p>
<p>It&rsquo;s also in use at Japanese executive job site <a href="http://ift.tt/2oSqQ2N" class="ext-link" rel="external nofollow">BizReach</a>, lifecycle management firm <a href="http://ift.tt/2BM7l1g" class="ext-link" rel="external nofollow">LiftIQ</a>, online developer training company&nbsp;<a href="http://ift.tt/1uvNK8t" class="ext-link" rel="external nofollow">Pluralsight</a>, and machine learning consultancy <a href="http://actionml.com/" class="ext-link" rel="external nofollow">ActionML</a>.</p>
<p></p>
<p>One of the project&rsquo;s biggest assets is a concept called <a href="http://ift.tt/2i4OTVB" class="ext-link" rel="external nofollow">template gallery</a>. It allows users to share their projects as a template. Companies building a similar application can simply download your template and run it as is or modify some of the components, said Chan, now senior director of product management for Salesforce Einstein.</p>
<p>The templates include a recommendation, classification, natural language processing and other tools. Regression templates include forecasting energy use, electric load and Boston housing prices.</p>
<p>Building a recommendation engine, a project that typically could take a team of team of expert data scientists months to do can be accomplished in a couple of weeks with one or two engineers with PredictionIO, he said in a <a href="http://ift.tt/2ldNJMk" class="ext-link" rel="external nofollow">blog post</a>.</p>
<h2>More than Algorithms</h2>
<p>PredictionIO originally was built on Hadoop, but has switched to <a href="http://ift.tt/1i4vP6x" class="ext-link" rel="external nofollow">Spark</a> as it has become more mature, Chan explained. It uses HBase as the data store; the Spark ML library <a href="http://ift.tt/1imGGJt" class="ext-link" rel="external nofollow">MLlib</a>; <a href="http://ift.tt/1ayLWPQ" class="ext-link" rel="external nofollow">Spray</a>, the JSON implementation in Scala; HDFS in some models and Elasticsearch to store metadata.</p>
<p>An event server continuously collects data from your application, in real-time or in batch, for model training and evaluation. The event server also can unify data from multiple sources for analysis.</p>
<p>The PredictionIO&nbsp;engine then builds models with one or more algorithms using the data. After it is deployed as a web service, it listens to queries from the application and responds through a REST API with predicted results in real-time.</p>
<p>While ASF has machine learning libraries such as <a href="http://ift.tt/2yKm5v8" class="ext-link" rel="external nofollow">Mahout</a> and <a href="http://ift.tt/1imGGJt" class="ext-link" rel="external nofollow">MLlib</a>, machine learning remains in early stages and needs researchers, developers and companies to contribute to it to make it successful, according to Chan.</p>
<p>&ldquo;If you&rsquo;re not really [knowledgeable] about the machine learning stack, you can&rsquo;t really do much with just algorithms,&rdquo; Chan said. There&rsquo;s a whole pipeline that PredictionIO handles, including preparing data, building algorithms, training, managing and evaluating models.</p>
<p><a href="http://ift.tt/1MCP4R0" class="ext-link" rel="external nofollow">SystemML</a>, another ASF project created at IBM, was also built on Spark and MLlib. It provides the ability to <a href="http://ift.tt/2rsqtJY" class="local-link">scale</a> data analysis from a small laptop to large clusters without the need to rewrite the entire codebase.</p>
<p>Other open source machine learning projects, such as&nbsp;<a href="http://ift.tt/1Xwlwg0" class="ext-link" rel="external nofollow">TensorFlow</a>,&nbsp; <a href="http://ift.tt/1gG1byZ" class="ext-link" rel="external nofollow">Caffe</a>,&nbsp;<a href="http://ift.tt/NdRQip" class="ext-link" rel="external nofollow">Vowpal Wabbit</a>, and&nbsp;<a href="http://ift.tt/O5hS6h" class="ext-link" rel="external nofollow">scikit-learn</a>, are ML libraries.</p>
<p>PredictionIO has been extended to support any JDBC (Java Database Connectivity) data store and multiple other algorithms, Chan said.</p>
<p>He sees more inter-project collaboration with ASF projects such as Spark and <a href="http://ift.tt/2iZJam5" class="ext-link" rel="external nofollow">Beam</a> as a result of the top-level status.</p>
<p>&ldquo;Usability is still a big problem in machine learning. We need to make the development process even easier, and that&rsquo;s going to require a lot of resources from the open source community,&rdquo; he said.</p>
<p>It&rsquo;s been encouraging to see how companies have been using both the open source and internal Salesforce implementations of PredictionIO, Chan said.</p>
<p>At the recent Dreamforce &rsquo;17 conference, for instance, staff from Ulster Bank in Scotland <a href="http://ift.tt/2oWGzOn" class="ext-link" rel="external nofollow">talked</a> about how it uses PredictionIO and Einstein for their <a href="http://ift.tt/2zm7cwt" class="ext-link" rel="external nofollow">Next Best Offer</a> predictions &mdash; using past customer behavior to know which banking services to suggest next.</p>
<p>And the University of Palermo, Argentina, has used PredictionIO to be able to predict <a href="http://ift.tt/2oWGB8X" class="ext-link" rel="external nofollow">dropout rates</a> for specific courses.</p>
<p class="attribution">Feature Image: &ldquo;<a href="http://ift.tt/2BQnHFQ" class="ext-link" rel="external nofollow">40+37 Magic 8 Ball</a>&rdquo; by Bark, licensed under&nbsp;<a href="http://ift.tt/N3rZKX" class="ext-link" rel="external nofollow">CC BY-SA 2.0</a>.</p>
<p>The post <a rel="nofollow" href="http://ift.tt/2CM6OsH">The PredictionIO Machine Learning Project Propels Intelligent Development</a> appeared first on <a rel="nofollow" href="https://thenewstack.io">The New Stack</a>.</p>
<br>
<br>
via Blog &ndash; The New Stack https://thenewstack.io<br>
December 20, 2017 at 06:00AM
